game/unicode.cc

```diff
@@ -37,62 +39,123 @@ std::string UnicodeUtil::getCharset (std::string const& str) {
 	return MimeEncodingName(encoding);
 }
 
-void UnicodeUtil::convertToUTF8 (std::stringstream &_stream, std::string _filename) {
-	std::string data = _stream.str();
-	std::string charset = UnicodeUtil::getCharset(data);	
+std::string UnicodeUtil::convertToUTF8 (std::string const& str, std::string _filename, CaseMapping toCase, bool assumeUTF8) {
+	std::string ret = str;
 	icu::UnicodeString ustring;
-	// Test for UTF-8 BOM (a three-byte sequence at the beginning of a file)
-	if (data.substr(0, 3) == "\xEF\xBB\xBF") {
-		_stream.str(data.substr(3)); // Remove BOM if there is one
+	std::string charset;
+	if (assumeUTF8) {
+		ustring = icu::UnicodeString::fromUTF8(str);
 	}
-	if (charset != "UTF-8") {
-		if (!_filename.empty()) { std::clog << "unicode/info: " << _filename << " does not appear to be UTF-8; (" << charset << ") detected." << std::endl; }
-		std::string _str;
-		const char* tmp = data.c_str();
-		icu::UnicodeString ustring = icu::UnicodeString(tmp, charset.c_str());
-		if (!ustring.isEmpty()) {
-			_stream.str(ustring.toUTF8String(_str));
+	else {
+		if (ret.substr(0, 3) == "\xEF\xBB\xBF") {
+			ret = ret.substr(3); // Remove BOM if there is one
+			charset = "UTF-8";
 		}
-		else {
-			if (!data.empty()) {
-				std::clog << "unicode/error: tried to convert text in an unknown encoding: " << charset << std::endl;
-			}
+		else { charset = UnicodeUtil::getCharset(str); }
+		// Test for UTF-8 BOM (a three-byte sequence at the beginning of a file)
+		if (charset != "UTF-8") {
+			if (!_filename.empty()) { std::clog << "unicode/info: " << _filename << " does not appear to be UTF-8; (" << charset << ") detected." << std::endl; }
+			ustring = icu::UnicodeString(ret.c_str(), charset.c_str());
+		}
+		else { ustring = icu::UnicodeString::fromUTF8(ret); }
+	}
+	switch(toCase) {
+		case CaseMapping::UPPER:
+			ustring.toUpper();
+			break;
+		case CaseMapping::LOWER:
+			ustring.toLower();
+			break;
+		case CaseMapping::TITLE:
+			ustring.toTitle(0, icu::Locale(Game::getSingletonPtr()->getCurrentLanguageCode().c_str()), U_TITLECASE_NO_LOWERCASE);
+			break;
+		case CaseMapping::NONE:
+			break;
+	}
+	if (!ustring.isEmpty()) {
+		ret.clear();
+		ustring.toUTF8String(ret);
+	}
+	else {
+		if (!ret.empty()) {
+			std::clog << "unicode/error: tried to convert text in an unknown encoding: " << charset << std::endl;
 		}
 	}
+	return ret;
+}
+
+bool UnicodeUtil::caseEqual (std::string const& lhs, std::string const& rhs, bool assumeUTF8) {
+	if (lhs == rhs) return true; // Early return
+	
+	std::string a = lhs;
+	std::string b = rhs;
+	std::string aCharset;
+	std::string bCharset;
```

Comment:
*Charset is used as flag here. So we should use bool and a const std::string for "UTF-8".

Reply:
No, because there are other valid values for the charset (essentially all old-timey code-pages, as well as Shift-JIS and such) which, if obtained, are then passed to the `UnicodeString` constructor.

Reply:
But where are the other values? There may be others but we use only this one. So why not using a static const? Even we have to use other values in the future they could also be static const. Actually I don't see that this must be dynamic.

Reply:
The values are returned by `getCharset`and passed directly to the `UnicodeString` constructor.
We check specifically for UTF-8 because there is a different factory method when you know for sure that you're dealing with UTF-8.